webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/AnnotationUserStateUtils.scala

```diff
@@ -0,0 +1,349 @@
+      val ownerSegmentVisibilityMap = id64BoolsToMap(ownerUserState.map(_.segmentVisibilities))
+      val mergedBoundingBoxVisibilityMap = ownerBoundingBoxVisibilityMap ++ requestingUserBoundingBoxVisibilityMap
+      val mergedSegmentGroupExpandedMap = ownerSegmentGroupExpandedMap ++ requestingUserSegmentGroupExpandedMap
+      val mergedSegmentVisibilityMap = ownerSegmentVisibilityMap ++ requestingUserSegmentVisibilityMap
+      VolumeUserStateProto(
+        userId = requestingUserId,
+        activeSegmentId =
+          requestingUserState.flatMap(_.activeSegmentId).orElse(ownerUserState.flatMap(_.activeSegmentId)),
+        segmentGroupExpandedStates = mapToId32Bools(mergedSegmentGroupExpandedMap),
+        boundingBoxVisibilities = mapToId32Bools(mergedBoundingBoxVisibilityMap),
+        segmentVisibilities = mapToId64Bools(mergedSegmentVisibilityMap)
+      )
+    }
+  }
+
+  private def mapId32Bools(idToBools: Seq[Id32ToBool], idMap: Map[Int, Int]): Seq[Id32ToBool] =
+    idToBools.map(mapIdBool(_, idMap))
+
+  private def mapIdBool(idToBools: Id32ToBool, idMap: Map[Int, Int]): Id32ToBool =
+    idToBools.copy(id = idMap.getOrElse(idToBools.id, idToBools.id))
+
+  private def mapId64Bools(idToBools: Seq[Id64ToBool], idMap: Map[Long, Long]): Seq[Id64ToBool] =
+    idToBools.map(mapIdBool(_, idMap))
+
+  private def mapIdBool(idToBools: Id64ToBool, idMap: Map[Long, Long]): Id64ToBool =
+    idToBools.copy(id = idMap.getOrElse(idToBools.id, idToBools.id))
+
+  private def mapIdBools(idToBools: Seq[Id32ToBool], functionalIdMapping: FunctionalGroupMapping): Seq[Id32ToBool] =
+    idToBools.map(mapIdBool(_, functionalIdMapping))
+
+  private def mapIdBool(idToBools: Id32ToBool, functionalIdMapping: FunctionalGroupMapping): Id32ToBool =
+    idToBools.copy(id = functionalIdMapping(idToBools.id))
+
+  // Merges user states of multiple skeleton tracings, respecting mapped ids of the tracing elements. The user set is preserved
+  protected def mergeSkeletonUserStates(tracingAUserStates: Seq[SkeletonUserStateProto],
+                                        tracingBUserStates: Seq[SkeletonUserStateProto],
+                                        groupMapping: FunctionalGroupMapping,
+                                        treeIdMapA: TreeIdMap,
+                                        treeIdMapB: TreeIdMap,
+                                        bboxIdMapA: UserBboxIdMap,
+                                        bboxIdMapB: UserBboxIdMap): Seq[SkeletonUserStateProto] = {
+    val tracingAUserStatesMapped =
+      tracingAUserStates.map(applyIdMappingsOnSkeletonUserState(_, groupMapping, treeIdMapA, bboxIdMapA))
+    val tracingBUserStatesMapped = tracingBUserStates
+      .map(userState => userState.copy(treeVisibilities = mapId32Bools(userState.treeVisibilities, treeIdMapB)))
+      .map(applyBboxIdMapOnSkeletonUserState(_, bboxIdMapB))
```

Comment:
If I remember correctly th were kept as they are and those of tracingA were increased to not collide with the onces from tracingB? Why is there a `treeIdMapB.getOrElse(treeId, treeId)` now :thinking:?

Reply:
The bounding box ids of both tracings may change, because there is a deduplication step (see the `distinctBy` in `BoundingBoxMerger`

Reply:
The deduplication step can mess with the bounding boxes of tracingB too. For example, if tracingB already has multiple boxes that have identical properties (except id), those would get deduplicated in the merge code, even if tracingA does not introduce new duplicates. So to be safe, I need to carry the id mapping for both A and B